include/Dialect/BGV/IR/BGVOps.td

```diff
@@ -0,0 +1,166 @@
+#ifndef HEIR_INCLUDE_DIALECT_BGV_IR_BGVOPS_TD_
+#define HEIR_INCLUDE_DIALECT_BGV_IR_BGVOPS_TD_
+
+include "include/Dialect/BGV/IR/BGVDialect.td"
+include "include/Dialect/BGV/IR/BGVTypes.td"
+include "BGVDialect.td"
+include "BGVTypes.td"
+include "mlir/Interfaces/InferTypeOpInterface.td"
+
+//===----------------------------------------------------------------------===//
+// BGV Operation definitions.
+//===----------------------------------------------------------------------===//
+
+class BGV_Op<string mnemonic, list<Trait> traits = []> :
+        Op<BGV_Dialect, mnemonic, traits> {
+
+  // See https://mlir.llvm.org/docs/DefiningDialects/Operations/#declarative-assembly-format
+  // TODO(b/291601096): Simplify this format by adding type constraints.
+  let assemblyFormat = [{
+    `(` operands `)` attr-dict `:`  `(` type(operands) `)` `->` type(results)
+  }];
+  let cppNamespace = "::mlir::heir::bgv";
+}
+
+//===----------------------------------------------------------------------===//
+// Homomorphic operations.
+//===----------------------------------------------------------------------===//
+
+def BGV_AddOp : BGV_Op<"add", [Commutative, SameOperandsAndResultType]> {
+  let summary = "Addition operation between ciphertexts.";
+
+  let arguments = (ins
+    Ciphertext:$x,
+    Ciphertext:$y
+  );
+
+  let results = (outs
+    Ciphertext:$output
+  );
+
+   let assemblyFormat = "`(` operands `)` attr-dict `:` type($output)" ;
+}
+
+def BGV_SubOp : BGV_Op<"sub", [SameOperandsAndResultType]> {
+  let summary = "Subtraction operation between ciphertexts.";
+
+  let arguments = (ins
+    Ciphertext:$x,
+    Ciphertext:$y
+  );
+
+  let results = (outs
+    Ciphertext:$output
+  );
+
+   let assemblyFormat = "`(` operands `)` attr-dict `:` type($output)" ;
+}
+
+def BGV_MulOp : BGV_Op<"mul", [Commutative, SameOperandsAndResultType]> {
```

Comment:
Quick note: if you have the (1, s) vs (1, s, s^2) attached to the type, then you can't use SameOperandsAndResultType. The different attributes, I believe, make the types different in the context of this trait.

Reply:
Ah good point! I think it's useful then to drop the attribute - mul would change the ciphertext attributes.